lib/AST/ASTPrinter.cpp

```diff
@@ -2389,6 +2389,32 @@ void PrintAST::visitStructDecl(StructDecl *decl) {
 
 void PrintAST::visitClassDecl(ClassDecl *decl) {
   printDocumentationComment(decl);
+
+  /// Determines if a class has any designated initializers that are
+  /// inaccessible to classes outside this module. If that's the case, then the
+  /// subclass cannot override all of this class's designated initializers and
+  /// therefore will not inherit the convenience initializers from the
+  /// superclass.
+  auto hasInaccessibleDesignatedInit = [&](ClassDecl *decl) -> bool {
+    auto members = decl->lookupDirect(DeclBaseName::createConstructor());
+    for (auto member : members) {
+      auto init = cast<ConstructorDecl>(member);
+      if (!init->isDesignatedInit())
+        continue;
+      if (!shouldPrint(init))
+        return true;
+    }
```

Comment:
Would this be clearer with `llvm::any_of`? You might not even need the lambda then.